### PR TITLE
Random port extended

### PR DESCRIPTION
**Pull request by petemiron**
### Changes proposed in this pull request: 

- Moves clustering and profiler to Random Ports
- Enables setting Random Ports in config files
- Adds capture of the last error string for future enhancements to unit tests
- Converts Default Option maps to methods generating a new map for each test
 
/cc @nats-io/core

### CODE REVIEW

File: server/reload_test.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ import (
 	"time"
 
 	"github.com/nats-io/go-nats"
+	"net"
```
Comment (tylertreat) [inline server/reload_test.go]: Nitpick: package import ordering should be:

1. stdlib
2. third-party
3. local

File: server/routes_test.go [hunk 2]
```diff
@@ -484,16 +494,18 @@ func TestRouteUseIPv6(t *testing.T) {
...
 	// Since client will first connect to the route listen port, set the
 	// cluster's Host to localhost so it works on Windows too, since on
 	// Windows, a client can't use 0.0.0.0 in a connect.
 	opts.Cluster.Host = "localhost"
 	opts.Cluster.NoAdvertise = true
-	s := RunServer(&opts)
-	defer s.Shutdown()
+	s := RunServer(opts)
+	s.Noticef("%+v\n", opts)
+	// defer s.Shutdown()
```
Comment (tylertreat) [inline server/routes_test.go]: Remove this?

File: server/server.go [hunk 3]
```diff
@@ -439,13 +447,38 @@ func (s *Server) StartProfiler() {
...
+	s.profiler = l
+	s.mu.Unlock()
+
 	go func() {
-		err := http.ListenAndServe(hp, nil)
-		if err != nil {
-			s.Fatalf("error starting monitor server: %s", err)
-		}
+		// if this errors out, it's probably because the server is being shutdown
+		srv.Serve(l)
```
Comment (tylertreat) [inline server/server.go]: Probably want to capture the error returned by `Serve`.

File: server/client_test.go [hunk 4]
```diff
@@ -634,20 +634,23 @@ func TestTwoTokenPubMatchSingleTokenSub(t *testing.T) {
...
-		t.Fatalf("Error creating client: %v\n", err)
+		t.Fatalf("Error creating client to %s: %v\n", url, err)
 	}
 	defer nc.Close()
 
 	ncp, err := nats.Connect(fmt.Sprintf("nats://%s:%d",
-		DefaultOptions.Host,
-		DefaultOptions.Port))
+		s.getOpts().Host,
+		s.Addr().(*net.TCPAddr).Port))
```
Comment (ColinSullivan1) [inline server/client_test.go]: A nitpick, but a convenience function to get the port GetPort(net.Addr) would seem a little cleaner throughout the tests.

Comment (petemiron) [inline server/client_test.go]: Resolved for new Addr() types (MonitorAddr and ClusterAddr). Left original Addr() as-is.

Comment (ColinSullivan1) [inline server/client_test.go]: Maybe `MonitorTCPAddr` and `ClusterTCPAddr` to differentiate the return types from the existing `Addr()`?

**Pull request merged**